Add automated testing and linting

chrisbsmith opened the pull request:
## What?

This PR adds automated testing and linting to ever PR that's submitted against a `release-*` branch. It leverages Helm's [chart testing](https://github.com/helm/chart-testing) tool to lint the charts. It lints all charts (individual services and the combo charts [fabric, sense, etc]). It also uses Gruntwork's [terratest](https://github.com/gruntwork-io/terratest/) to automate a deployment of Grey Matter into k3d using helm. (I had to fork their repo and add an update for `helm dep`)

## Why?

Why not?

## How?

In order to achieve this, I had to make a few modifications to the charts:

* all pods now have a `greymatter: <chart>` label. This is used by the helm tests to count the number of pods expected with each chart
* All charts were updated to pass the linter. Some big notes:
  * The Organization config in `Chart.yaml` is compared against GitHub. I changed this to our `greymatter-io` org in anticipation of moving the helm-charts there
  * The website config was removed
  * The `appVersion` has to be a three digit semver, so a `.0` was added to all that were to short
  * **The linter will fail if the chart version is not updated**

The linter and the tests should run whenever a PR is submitted to a `release-*` branch, and the packaging will only occur if the linter and testing pass.  

Definitely would love eyes on this @zoemccormick @tilleryd @dontlaugh @moosecanswim , though technically there is no code change.

dontlaugh commented on .circleci/old.config.yml:
```diff
@@ -0,0 +1,261 @@
+# These are the expected environment variables
```
We don't need to check this in, imo

chrisbsmith replied:
ha, good catch.

chrisbsmith commented:
@dontlaugh I meant to add the slack orb to this.  I'll look into this weekend, otherwise I'll put another PR up later if this one makes it in first.  If you know how to do it, I'll take any help

dontlaugh commented on test/greymatter_integration_test.go:
```diff
@@ -0,0 +1,277 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/chrisbsmith/terratest/modules/helm"
```
is this a fork?

chrisbsmith replied:
Yes, but with my `dependency` module added.

chrisbsmith replied:
https://github.com/chrisbsmith/terratest/blob/master/modules/helm/dependency.go

I need to submit it back to the upstream, but things get busy

dontlaugh replied:
Gotcha, try and get that upstreamed. Otherwise, I'd just wrap the shelling-out right here in this file for `helm dep`.

dontlaugh reviewed (commented):
looks good, some questions and comments

dontlaugh commented:
I've never used an orb, fwiw :shrug:

dontlaugh commented on test/go.mod:
```diff
@@ -0,0 +1,11 @@
+module github.com/chrisbsmith/terratest-helm-testing-example/test
```
_this_ module path should probably be re-initialized to github.com/DecipherNow/helm-charts

dontlaugh reviewed (commented):
LGTM like this but I'd like another set of eyes 

:+1:

chrisbsmith commented:
@dontlaugh , I believe I addressed all of your concerns.  I've removed the need to use my fork and I've added slack integration so it notifies the #decipher-sres-bots channel.  

Please double check my branch filtering logic in the workflow though.

zoemccormick reviewed (approved):
this looks great to me! Everything comes up as expected in the test and in regular behavior.

dontlaugh commented:
This branch filtering logic looks good to me
```yaml
          branches:
              only: /release-.*/
```

moosecanswim commented on test/greymatter_integration_test.go:
```diff
@@ -0,0 +1,277 @@
+		&tlsConfig,
+		10,
+		6*time.Second,
+		func(statusCode int, body string) bool {
```
You may want to add a statusCode check before trying trying to unmarshal a response.  If the response is bad (maybe a 503) then `body` will be null which will throw an error at `json.Unmarshal`.  if a statusCode check is put in place then we can get emit an error that more accurately reflects whats going on.

chrisbsmith replied:
So this method `HttpGetWithRetryWithCustomValidation` uses the function to validate the response. It already uses the `statusCode` to catch a non-true value and then retries for 10 times, waiting 6 second between tries.

chrisbsmith replied:
https://github.com/gruntwork-io/terratest/blob/0c509a63364eaee678570c2f64c254c850b2ef19/modules/http-helper/http_helper.go#L78

moosecanswim commented on test/greymatter_integration_test.go:
```diff
@@ -0,0 +1,277 @@
+	}
+}
+
+func verifyCatalog(t *testing.T, kubectlOptions *k8s.KubectlOptions) {
```
May want to add a retry similar to how the kubernetes resource checks are done

chrisbsmith replied:
The retry logic is actually included in the `HttpGetWithRetryWithCustomValidation` function.

chrisbsmith commented:
Closes #730 
Closes #731 
Closes #732 
Closes #733

moosecanswim reviewed (approved):
LGTM thanks for the clarification on the retry and status checks for the catalog tests!

chrisbsmith commented:
🤞
